File: labs/20-human-in-the-loop/python/main.py
"""Human-in-the-loop workflow.

Pauses for human approval before executing a destructive action and supports
resuming from persisted state.
"""
import argparse
import json
import logging
import sys
import time
from pathlib import Path
from typing import Any

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "shared" / "config"))

from openai_client import configure_logging

logger = logging.getLogger("human-in-the-loop")

STATE_FILE = Path(__file__).with_suffix("").parent / "workflow_state.json"


class DestructiveAction:
    def __init__(self, target: str, action: str) -> None:
        self.target = target
        self.action = action

    def describe(self) -> str:
        return f"{self.action} {self.target}"


class HumanInTheLoopWorkflow:
    def __init__(self, state_file: Path) -> None:
        self.state_file = state_file
        self.state = self._load()

    def _load(self) -> dict[str, Any]:
        if self.state_file.exists():
            with open(self.state_file, "r", encoding="utf-8") as f:
                return json.load(f)
        return {}

    def _save(self) -> None:
        tmp = self.state_file.with_suffix(".tmp")
        with open(tmp, "w", encoding="utf-8") as f:
            json.dump(self.state, f, indent=2)
        tmp.replace(self.state_file)

    def reset(self) -> None:
        if self.state_file.exists():
            self.state_file.unlink()
        self.state = {}

    def request_approval(self, action: DestructiveAction) -> None:
        if self.state.get("stage") in ("approved", "executed", "rejected"):
            return
        print(f"\nDestructive action requested: {action.describe()}")
        print("This action cannot be undone. Please review carefully.")
        self.state = {
            "stage": "awaiting_approval",
            "target": action.target,
            "action": action.action,
            "approved": None,
        }
        self._save()

    def collect_decision(self) -> bool | None:
        if self.state.get("stage") != "awaiting_approval":
            return self.state.get("approved")

        while True:
            answer = input("Approve? [y/n]: ").strip().lower()
            if answer in ("y", "yes"):
                self.state["approved"] = True
                self.state["stage"] = "approved"
                self._save()
                return True
            if answer in ("n", "no"):
                self.state["approved"] = False
                self.state["stage"] = "rejected"
                self._save()
                return False
            print("Please answer 'y' or 'n'.")

    def execute(self, action: DestructiveAction) -> None:
        if self.state.get("stage") == "executed":
            print("Action already executed.")
            return
        if not self.state.get("approved"):
            print("Action not approved; will not execute.")
            return
        print(f"\nExecuting: {action.describe()}")
        # Simulate destructive work.
        print(f"  -> {action.target} has been {action.action}d.")
        self.state["stage"] = "executed"
        self.state["executed_at"] = time.strftime("%Y-%m-%dT%H:%M:%S")
        self._save()

    def run(self, action: DestructiveAction) -> None:
        self.request_approval(action)
        approved = self.collect_decision()
        if approved:
            self.execute(action)
        else:
            print("\nAction rejected by human operator.")


def main() -> None:
    configure_logging()
    parser = argparse.ArgumentParser(description="Human-in-the-loop destructive action")
    parser.add_argument("--reset", action="store_true", help="Clear saved workflow state")
    args = parser.parse_args()

    workflow = HumanInTheLoopWorkflow(STATE_FILE)
    if args.reset:
        workflow.reset()
        print("Workflow state reset.")
        return

    action = DestructiveAction(target="user_account_42", action="delete")
    workflow.run(action)


if __name__ == "__main__":
    main()

File: labs/20-human-in-the-loop/node/main.ts
import fs from "fs";
import path from "path";
import readline from "readline";

const STATE_FILE = path.join(__dirname, "workflow_state.json");

class DestructiveAction {
  constructor(public target: string, public action: string) {}

  describe(): string {
    return `${this.action} ${this.target}`;
  }
}

class HumanInTheLoopWorkflow {
  private state: any;

  constructor(private stateFile: string) {
    this.state = this.load();
  }

  private load(): any {
    if (fs.existsSync(this.stateFile)) {
      return JSON.parse(fs.readFileSync(this.stateFile, "utf8"));
    }
    return {};
  }

  private save(): void {
    const tmp = `${this.stateFile}.tmp`;
    fs.writeFileSync(tmp, JSON.stringify(this.state, null, 2), "utf8");
    fs.renameSync(tmp, this.stateFile);
  }

  reset(): void {
    if (fs.existsSync(this.stateFile)) fs.unlinkSync(this.stateFile);
    this.state = {};
  }

  requestApproval(action: DestructiveAction): void {
    if (["approved", "executed", "rejected"].includes(this.state.stage)) return;
    console.log(`\nDestructive action requested: ${action.describe()}`);
    console.log("This action cannot be undone. Please review carefully.");
    this.state = {
      stage: "awaiting_approval",
      target: action.target,
      action: action.action,
      approved: null,
    };
    this.save();
  }

  async collectDecision(): Promise<boolean | null> {
    if (this.state.stage !== "awaiting_approval") {
      return this.state.approved;
    }
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    const ask = (q: string): Promise<string> => new Promise((resolve) => rl.question(q, resolve));

    while (true) {
      const answer = (await ask("Approve? [y/n]: ")).trim().toLowerCase();
      if (answer === "y" || answer === "yes") {
        this.state.approved = true;
        this.state.stage = "approved";
        this.save();
        rl.close();
        return true;
      }
      if (answer === "n" || answer === "no") {
        this.state.approved = false;
        this.state.stage = "rejected";
        this.save();
        rl.close();
        return false;
      }
      console.log("Please answer 'y' or 'n'.");
    }
  }

  execute(action: DestructiveAction): void {
    if (this.state.stage === "executed") {
      console.log("Action already executed.");
      return;
    }
    if (!this.state.approved) {
      console.log("Action not approved; will not execute.");
      return;
    }
    console.log(`\nExecuting: ${action.describe()}`);
    console.log(`  -> ${action.target} has been ${action.action}d.`);
    this.state.stage = "executed";
    this.state.executed_at = new Date().toISOString();
    this.save();
  }

  async run(action: DestructiveAction): Promise<void> {
    this.requestApproval(action);
    const approved = await this.collectDecision();
    if (approved) {
      this.execute(action);
    } else {
      console.log("\nAction rejected by human operator.");
    }
  }
}

async function main() {
  const reset = process.argv.includes("--reset");
  const workflow = new HumanInTheLoopWorkflow(STATE_FILE);
  if (reset) {
    workflow.reset();
    console.log("Workflow state reset.");
    return;
  }

  const action = new DestructiveAction("user_account_42", "delete");
  await workflow.run(action);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
